src/main/java/com/microsoft/sqlserver/jdbc/SQLServerXAResource.java

```diff
@@ -751,9 +752,24 @@ private XAReturnValue DTC_XA_Interface(int nType, Xid xid, int xaFlags) throws X
                     }
                 }
             }
-        } catch (SQLServerException | SQLTimeoutException ex) {
+        } catch (SQLTimeoutException ex) {
             if (xaLogger.isLoggable(Level.FINER))
                 xaLogger.finer(toString() + " exception:" + ex);
+            XAException e = new XAException(ex.toString());
+            e.errorCode = XAException.XAER_RMFAIL;
+            throw e;
+
+        } catch (SQLServerException ex) {
+            if (xaLogger.isLoggable(Level.FINER))
+                xaLogger.finer(toString() + " exception:" + ex);
+
+            if (ex.getMessage().equals(SQLServerException.getErrString("R_noServerResponse"))
+                    || TransientError.isTransientError(ex.getSQLServerError())) {
+                XAException e = new XAException(ex.toString());
+                e.errorCode = XAException.XAER_RMFAIL;
+                throw e;
+            }
+
```

Comment:
this makes me wonder about other errors that we should not RMERR on. I see similar issues in Oracle as well. I'm wondering if we should check severity?

Reply:
As discussed offline, we expect the statement we call to return an XA status (including any XA errors). Any other error the driver receives, like a server-level error, it would be difficult to discern/translate those into the correct RMERR/RMFAIL status. So going with the conservative change here and only changing the specific behavior we believe should be changed from RMERR to RMFAIL. Other cases will remain the same (RMERR).